---
title: "Differential Expression"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/RNAseq/DE.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data

```{r libraries, message=FALSE}
library(DESeq2)
library(openxlsx)
library(goseq)
library(GO.db)
library(org.Mm.eg.db)
```

```{r, data}
data.all <- read.table("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Data/FilteredCounts.txt", sep = "\t")
data.all <- data.all[rowMeans(data.all) > 1,]

genelength <- read.table("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Data/Genelength.txt", header = TRUE, sep = "\t")

groups <- paste(sapply(colnames(data.all), 
                                    function(n){unlist(strsplit(n, "_"))[1]}),
                sapply(colnames(data.all), 
                                    function(n){unlist(strsplit(n, "_"))[2]}),
                sapply(colnames(data.all), 
                                    function(n){unlist(strsplit(n, "_"))[3]}),
                sep = "_")

# Read in genenames
genes <- read.table("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genes) <- genes$Gene.stable.ID

# List of pairwise comparisons
list.comp <- list(#"Vg1_Th1_young vs Vg4_Th1_young" = c("Vg1_Th1_young", "Vg4_Th1_young"),
                  #"Vg4_Th17_young vs Vg6_Th17_young" = c("Vg4_Th17_young", "Vg6_Th17_young"),
                  "Vg4_Th1_young vs Vg4_Th17_young" = c("Vg4_Th1_young", "Vg4_Th17_young"),
                  "Vg1_Th1_young vs Vg1_Th1_old" = c("Vg1_Th1_young", "Vg1_Th1_old"),
                  "Vg4_Th1_young vs Vg4_Th1_old" = c("Vg4_Th1_young", "Vg4_Th1_old"),
                  "Vg4_Th17_young vs Vg4_Th17_old" = c("Vg4_Th17_young", "Vg4_Th17_old"),
                  "Vg6_Th17_young vs Vg6_Th17_old" = c("Vg6_Th17_young", "Vg6_Th17_old"),
                  #"Vg1_Th1_old vs Vg4_Th1_old" = c("Vg1_Th1_old", "Vg4_Th1_old"),
                  #"Vg4_Th17_old vs Vg6_Th17_old" = c("Vg4_Th17_old", "Vg6_Th17_old"),
                  "Vg4_Th1_old vs Vg4_Th17_old" = c("Vg4_Th1_old", "Vg4_Th17_old"))
```

# Differential expression analysis

```{r}
for(i in 1:length(list.comp)){
  n <- list.comp[[i]]
  cur_data <- data.all[,grepl(n[1], colnames(data.all)) |
                                                           grepl(n[2], colnames(data.all))]
  cur_data <- cur_data[rowMeans(cur_data) > 0,]
  cur_groups <- groups[grepl(n[1], colnames(data.all)) |
                                                           grepl(n[2], colnames(data.all))]
  cur_names <- colnames(data.all)[grepl(n[1], colnames(data.all)) |
                                                           grepl(n[2], colnames(data.all))]
  
  cur_dds <- DESeqDataSetFromMatrix(countData = cur_data, 
        colData = data.frame(row.names = cur_names,
                                  groups = cur_groups),
                              design = ~ groups)
  cur_dds <- DESeq(cur_dds)
  cur_res_all <- results(cur_dds)
  cur_res_all <- cur_res_all[!is.na(cur_res_all$padj),]
  cur_res <- cur_res_all[cur_res_all$padj < 0.1,]
  cur_res$Genename <- genes[rownames(cur_res),2]
  cur_res.1 <- cur_res[cur_res$log2FoldChange < 0,]
  cur_res.1 <- cur_res.1[order(cur_res.1$padj, decreasing = FALSE),]
  cur_res.1$GeneID <- rownames(cur_res.1)
  cur_res.2 <- cur_res[cur_res$log2FoldChange > 0,]
  cur_res.2 <- cur_res.2[order(cur_res.2$padj, decreasing = FALSE),] 
  cur_res.2$GeneID <- rownames(cur_res.2)
  cur_list <- list()
  cur_list[[levels(cur_dds$groups)[1]]] <- cur_res.1
  cur_list[[levels(cur_dds$groups)[2]]] <- cur_res.2
  
  # GO analysis
  cur_list.go <- list()
  
  cur_genes <- as.integer(cur_res_all$padj < 0.1 & cur_res_all$log2FoldChange < 0)
  names(cur_genes) <- rownames(cur_res_all)
  
  if(sum(cur_genes == 1) < 5){
    enriched.GO <- NULL
  }
  else{
      pwf=nullp(cur_genes,"mm10","ensGene", bias.data = genelength[names(cur_genes),])
      GO.wall=goseq(pwf,"mm10","ensGene")
      enriched.GO=GO.wall[p.adjust(GO.wall$over_represented_pvalue,method="fdr")<.1,]
  }
  
  # Add genenames to the GO categories
  if(nrow(enriched.GO) > 0){
    all_genes <- vector(length = nrow(enriched.GO))
    for(j in 1:nrow(enriched.GO)){
      allegs = get(enriched.GO$category[j], org.Mm.egGO2ALLEGS)
      genes.GO = unlist(mget(allegs,org.Mm.egSYMBOL))
      genes.GO = as.character(genes.GO[genes.GO %in% cur_res.1$Genename])
      all_genes[j] <- paste(genes.GO, collapse = ", ")
    }
    enriched.GO$Genes <- all_genes
  }
  
  cur_list.go[[levels(cur_dds$groups)[1]]] <- enriched.GO
  
  cur_genes <- as.integer(cur_res_all$padj < 0.1 & cur_res_all$log2FoldChange > 0)
  names(cur_genes) <- rownames(cur_res_all)
  
  if(sum(cur_genes == 1) < 5){
    enriched.GO <- NULL
  }
  else{
      pwf=nullp(cur_genes,"mm10","ensGene", bias.data = genelength[names(cur_genes),])
      GO.wall=goseq(pwf,"mm10","ensGene")
      enriched.GO=GO.wall[p.adjust(GO.wall$over_represented_pvalue,method="fdr")<.1,]
  }
  # Add genenames to the GO categories 
  if(nrow(enriched.GO) > 0){
    all_genes <- vector(length = nrow(enriched.GO))
    for(j in 1:nrow(enriched.GO)){
      allegs = get(enriched.GO$category[j], org.Mm.egGO2ALLEGS)
      genes.GO = unlist(mget(allegs,org.Mm.egSYMBOL))
      genes.GO = as.character(genes.GO[genes.GO %in% cur_res.2$Genename])
      all_genes[j] <- paste(genes.GO, collapse = ", ")
    }
    enriched.GO$Genes <- all_genes
  }
  
  cur_list.go[[levels(cur_dds$groups)[2]]] <- enriched.GO
  
  write.xlsx(cur_list, paste("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Marker_genes/", names(list.comp)[i], ".xlsx", sep = ""))
  write.xlsx(cur_list.go, paste("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Marker_genes/", names(list.comp)[i], "_GO.xlsx", sep = ""))
}
```