---
title: "Differential Expression"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/RNAseq/DE.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data

```{r libraries, message=FALSE}
library(edgeR)
library(openxlsx)
library(goseq)
library(GO.db)
library(org.Mm.eg.db)
library(VennDiagram)
library(pheatmap)
library(viridis)
library(gsubfn)
```

```{r, data}
data.all <- read.table("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Data/FilteredCounts.txt", sep = "\t")
data.all <- data.all[rowMeans(data.all) > 1,]

genelength <- read.table("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Data/Genelength.txt", header = TRUE, sep = "\t")

groups <- paste(sapply(colnames(data.all), 
                                    function(n){unlist(strsplit(n, "_"))[1]}),
                sapply(colnames(data.all), 
                                    function(n){unlist(strsplit(n, "_"))[2]}),
                sapply(colnames(data.all), 
                                    function(n){unlist(strsplit(n, "_"))[3]}),
                sep = "_")

# Read in genenames
genes <- read.table("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genes) <- genes$Gene.stable.ID

# List of pairwise comparisons
list.comp <- list("Vg1_Th1_young vs Vg4_Th1_young" = c("Vg1_Th1_young", "Vg4_Th1_young"),
                  "Vg4_Th17_young vs Vg6_Th17_young" = c("Vg4_Th17_young", "Vg6_Th17_young"),
                  "Vg4_Th1_young vs Vg4_Th17_young" = c("Vg4_Th1_young", "Vg4_Th17_young"),
                  "Vg1_Th1_young vs Vg1_Th1_old" = c("Vg1_Th1_young", "Vg1_Th1_old"),
                  "Vg4_Th1_young vs Vg4_Th1_old" = c("Vg4_Th1_young", "Vg4_Th1_old"),
                  "Vg4_Th17_young vs Vg4_Th17_old" = c("Vg4_Th17_young", "Vg4_Th17_old"),
                  "Vg6_Th17_young vs Vg6_Th17_old" = c("Vg6_Th17_young", "Vg6_Th17_old"),
                  "Vg1_Th1_old vs Vg4_Th1_old" = c("Vg1_Th1_old", "Vg4_Th1_old"),
                  "Vg4_Th17_old vs Vg6_Th17_old" = c("Vg4_Th17_old", "Vg6_Th17_old"),
                  "Vg4_Th1_old vs Vg4_Th17_old" = c("Vg4_Th1_old", "Vg4_Th17_old"),
                  "Th1 vs Th17" = c("Th1_", "Th17_")
                  )
```

# Differential expression analysis

The differential expression analysis is performed between the groups listed above.

```{r}
for(i in 1:length(list.comp)){
  n <- list.comp[[i]]
  cur_data <- data.all[,grepl(n[1], colnames(data.all)) |
                              grepl(n[2], colnames(data.all))]
  cur_data <- cur_data[rowMeans(cur_data) > 10,]
  
  cur_groups <- unlist(strapplyc(groups, paste(n[1], n[2], sep = "|")))
  cur_names <- colnames(data.all)[grepl(n[1], colnames(data.all)) |
                          grepl(n[2], colnames(data.all))]
  
  # Differential expression testing using edgeR
  y <- DGEList(counts=cur_data,
             group=cur_groups)
  y <- calcNormFactors(y)
  design <- model.matrix(~0+y$samples$group)
  colnames(design) <- sub("y\\$samples\\$group", "", colnames(design))
  y <- estimateDisp(y,design)
  
  fit <- glmQLFit(y,design, robust = TRUE)
  qlf <- glmQLFTest(fit,coef=2, 
                  contrast = makeContrasts(contrasts = paste(colnames(design)[2], 
                                            "-", colnames(design)[1]), 
                                           levels = design))
  cur_res_all <- topTags(qlf, n = nrow(qlf$table))$table

  cur_res_all <- cur_res_all[!is.na(cur_res_all$FDR),]
  cur_res <- cur_res_all[cur_res_all$FDR < 0.1,]
  cur_res$Genename <- genes[rownames(cur_res),2]
  cur_res.1 <- cur_res[cur_res$logFC < 0,]
  cur_res.1 <- cur_res.1[order(cur_res.1$FDR, decreasing = FALSE),]
  cur_res.1$GeneID <- rownames(cur_res.1)
  cur_res.2 <- cur_res[cur_res$logFC > 0,]
  cur_res.2 <- cur_res.2[order(cur_res.2$FDR, decreasing = FALSE),] 
  cur_res.2$GeneID <- rownames(cur_res.2)
  cur_list <- list()
  cur_list[[colnames(design)[1]]] <- cur_res.1
  cur_list[[colnames(design)[2]]] <- cur_res.2
  
  # GO analysis
  cur_list.go <- list()
  
  cur_genes <- as.integer(cur_res_all$FDR < 0.1 & cur_res_all$logFC < 0)
  names(cur_genes) <- rownames(cur_res_all)
  
  if(sum(cur_genes == 1) < 5){
    enriched.GO <- matrix(data = NA, nrow = 1, ncol = 5)
  }
  else{
      pwf=nullp(cur_genes,"mm10","ensGene", bias.data = genelength[names(cur_genes),])
      GO.wall=goseq(pwf,"mm10","ensGene")
      enriched.GO=GO.wall[p.adjust(GO.wall$over_represented_pvalue,method="fdr")<.1,]
  }
  
  # Add genenames to the GO categories
  if(sum(cur_genes == 1) > 5){
  if(nrow(enriched.GO) > 0){
    all_genes <- vector(length = nrow(enriched.GO))
    for(j in 1:nrow(enriched.GO)){
      allegs = get(enriched.GO$category[j], org.Mm.egGO2ALLEGS)
      genes.GO = unlist(mget(allegs,org.Mm.egSYMBOL))
      genes.GO = unique(as.character(genes.GO[genes.GO %in% cur_res.1$Genename]))
      all_genes[j] <- paste(genes.GO, collapse = ", ")
    }
    enriched.GO$Genes <- all_genes
  }
  }
  
  cur_list.go[[colnames(design)[1]]] <- enriched.GO
  
  cur_genes <- as.integer(cur_res_all$FDR < 0.1 & cur_res_all$logFC > 0)
  names(cur_genes) <- rownames(cur_res_all)
  
  if(sum(cur_genes == 1) < 5){
    enriched.GO <- matrix(data = NA, nrow = 1, ncol = 5)
  }
  else{
      pwf=nullp(cur_genes,"mm10","ensGene", bias.data = genelength[names(cur_genes),])
      GO.wall=goseq(pwf,"mm10","ensGene")
      enriched.GO=GO.wall[p.adjust(GO.wall$over_represented_pvalue,method="fdr")<.1,]
  }
  # Add genenames to the GO categories 
  if(sum(cur_genes == 1) > 5){
  if(nrow(enriched.GO) > 0){
    all_genes <- vector(length = nrow(enriched.GO))
    for(j in 1:nrow(enriched.GO)){
      allegs = get(enriched.GO$category[j], org.Mm.egGO2ALLEGS)
      genes.GO = unlist(mget(allegs,org.Mm.egSYMBOL))
      genes.GO = unique(as.character(genes.GO[genes.GO %in% cur_res.2$Genename]))
      all_genes[j] <- paste(genes.GO, collapse = ", ")
    }
    enriched.GO$Genes <- all_genes
  }
  }
  
  cur_list.go[[colnames(design)[2]]] <- enriched.GO
  
  write.xlsx(cur_list, paste("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Marker_genes_new/", names(list.comp)[i], ".xlsx", sep = ""))
  write.xlsx(cur_list.go, paste("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Marker_genes_new/", names(list.comp)[i], "_GO.xlsx", sep = ""))
}
```

# Comparison between Th1 and Th17

```{r Th1_vs_Th17}
# Read in relevant data
Th1young <- read.xlsx("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Marker_genes/Vg4_Th1_young vs Vg4_Th17_young.xlsx", sheet = 1)
Th17young <- read.xlsx("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Marker_genes/Vg4_Th1_young vs Vg4_Th17_young.xlsx", sheet = 2)

Th1old <- read.xlsx("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Marker_genes/Vg4_Th1_old vs Vg4_Th17_old.xlsx", sheet = 1)
Th17old <- read.xlsx("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Marker_genes/Vg4_Th1_old vs Vg4_Th17_old.xlsx", sheet = 2)

# Find the overlab between these lists
shared.Th1 <- intersect(Th1young$GeneID,
                        Th1old$GeneID)

shared.Th17 <- intersect(Th17young$GeneID,
                        Th17old$GeneID)

# VennDiagram for Th1
pdf("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Th1_Th17_comparison/VennDiagram_Th1.pdf",
    onefile = FALSE)
draw.pairwise.venn(area1 = length(Th1young$GeneID), 
                   area2 = length(Th1old$GeneID), 
                   cross.area = length(shared.Th1), 
                   category = c("Up in Th1 young", 
                                "Up in Th1 old"), 
                   euler.d = TRUE, scaled = TRUE,
                   fill = c("steelblue4", "coral3"))
dev.off()

# Save Th1 old signature
write.xlsx(Th1old[is.na(match(Th1old$GeneID, shared.Th1)),], 
           "/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Th1_Th17_comparison/Th1_old_signature.xlsx")
write.xlsx(Th17old[is.na(match(Th17old$GeneID, shared.Th17)),], 
           "/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Th1_Th17_comparison/Th17_old_signature.xlsx")

# VennDiagram for Th17
pdf("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Th1_Th17_comparison/VennDiagram_Th17.pdf",
    onefile = FALSE)
draw.pairwise.venn(area1 = length(Th17young$GeneID), 
                   area2 = length(Th17old$GeneID), 
                   cross.area = length(shared.Th17), 
                   category = c("Up in Th17 young", 
                                "Up in Th17 old"), 
                   euler.d = TRUE, scaled = TRUE,
                   fill = c("steelblue4", "coral3"))
dev.off()
```

Visualization of heatmaps.

```{r heatmap}
# Heatmaps - Th1 young
cur_heatmap <- data.all[Th1young$GeneID[1:100],
                        c(which(grepl("Vg4_Th1_young", colnames(data.all))),
                      which(grepl("Vg4_Th17_young", colnames(data.all))))]

pdf("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Th1_Th17_comparison/Heatmap_Th1young.pdf",
    onefile = FALSE, height = 15, width = 8)
pheatmap(log10(cur_heatmap + 1),
         cluster_rows = FALSE, cluster_cols = FALSE, color = viridis(100),
         labels_row = Th1young$Genename[1:100], border_color = NA, 
         show_colnames = FALSE, cellheight = 8, cellwidth = 8, fontsize = 7,
         annotation_col = data.frame(row.names = colnames(cur_heatmap),
                          group = sub("_DO.*$", "", colnames(cur_heatmap))))
dev.off()

# Heatmaps - Th17 young
cur_heatmap <- data.all[Th17young$GeneID[1:100],
                        c(which(grepl("Vg4_Th1_young", colnames(data.all))),
                      which(grepl("Vg4_Th17_young", colnames(data.all))))]

pdf("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Th1_Th17_comparison/Heatmap_Th17young.pdf",
    onefile = FALSE, height = 15, width = 8)
pheatmap(log10(cur_heatmap + 1),
         cluster_rows = FALSE, cluster_cols = FALSE, color = viridis(100),
         labels_row = Th17young$Genename[1:100], border_color = NA, 
         show_colnames = FALSE, cellheight = 8, cellwidth = 8, fontsize = 7,
         annotation_col = data.frame(row.names = colnames(cur_heatmap),
                          group = sub("_DO.*$", "", colnames(cur_heatmap))))
dev.off()

# Heatmaps - Th1 old
cur_heatmap <- data.all[Th1old$GeneID[1:100],
                        c(which(grepl("Vg4_Th1_old", colnames(data.all))),
                      which(grepl("Vg4_Th17_old", colnames(data.all))))]

pdf("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Th1_Th17_comparison/Heatmap_Th1old.pdf",
    onefile = FALSE, height = 15, width = 8)
pheatmap(log10(cur_heatmap + 1),
         cluster_rows = FALSE, cluster_cols = FALSE, color = viridis(100),
         labels_row = Th1old$Genename[1:100], border_color = NA, 
         show_colnames = FALSE, cellheight = 8, cellwidth = 8, fontsize = 7,
         annotation_col = data.frame(row.names = colnames(cur_heatmap),
                          group = sub("_DO.*$", "", colnames(cur_heatmap))))
dev.off()

# Heatmaps - Th17 old
cur_heatmap <- data.all[Th17old$GeneID[1:100],
                        c(which(grepl("Vg4_Th1_old", colnames(data.all))),
                      which(grepl("Vg4_Th17_old", colnames(data.all))))]

pdf("/Users/nils/Dropbox (Cambridge University)/GammaDelta/Analysis/Results/Th1_Th17_comparison/Heatmap_Th17old.pdf",
    onefile = FALSE, height = 15, width = 8)
pheatmap(log10(cur_heatmap + 1),
         cluster_rows = FALSE, cluster_cols = FALSE, color = viridis(100),
         labels_row = Th17old$Genename[1:100], border_color = NA, 
         show_colnames = FALSE, cellheight = 8, cellwidth = 8, fontsize = 7,
         annotation_col = data.frame(row.names = colnames(cur_heatmap),
                          group = sub("_DO.*$", "", colnames(cur_heatmap))))
dev.off()
```